contrib/TESTING.md

```diff
@@ -0,0 +1,95 @@
+# Testing Guide for Bittensor
+
+Testing is an essential part of software development that ensures the correctness and performance of your code. Bittensor uses a combination of unit tests and integration tests to verify the functionality of its components. This guide will walk you through how to run and write tests for Bittensor.
+
+## Running Tests
+
+Bittensor uses `pytest` for running its tests. To run all tests, navigate to the root directory of the Bittensor repository and run:
+
+```bash
+pytest
+```
+
+This will automatically discover all test files (those that start with `test_`) and run them.
+
+If you want to run a specific test file, you can specify it directly. For example, to run the tests in `test_wallet.py`, you would use:
+
+```bash
+pytest tests/test_wallet.py
+```
+
+Similarly, you can run a specific test within a file by appending `::` and the test name. For example:
+
+```bash
+pytest tests/test_wallet.py::test_create_new_coldkey
+```
+
+## Writing Tests
+
+When writing tests for Bittensor, you should aim to cover both the "happy path" (where everything works as expected) and any potential error conditions. Here's a basic structure for a test file:
+
+```python
+import pytest
+import bittensor
+
+def test_some_functionality():
+    # Setup any necessary objects or state.
+    wallet = bittensor.wallet()
+
+    # Call the function you're testing.
+    result = wallet.create_new_coldkey()
+
+    # Assert that the function behaved as expected.
+    assert result is not None
+```
+
+In this example, we're testing the `create_new_coldkey` function of the `wallet` object. We assert that the result is not `None`, which is the expected behavior.
+
+## Mocking
+
+In some cases, you may need to mock certain functions or objects to isolate the functionality you're testing. Bittensor uses the `unittest.mock` library for this. Here's an example:
+
+```python
+from unittest.mock import MagicMock
+
+def test_some_functionality():
+    # Create a mock object.
+    mock_wallet = MagicMock()
+
+    # Set the return value for a specific function.
+    mock_wallet.create_new_coldkey.return_value = "mocked_key"
+
+    # Call the function you're testing.
+    result = mock_wallet.create_new_coldkey()
+
+    # Assert that the function behaved as expected.
+    assert result == "mocked_key"
+```
```

Comment:
Focusing on the accuracy of the tests, is that actually testing something inside bittensor? or just proving that the MagicMock works?

Reply:
Good point.  I will get a real example of MagicMock, as this one is misleading.

Reply:
See: cfaf66b8d500383df80bdaf7ac0f4f790a5489c9